viv_utils/flirt.py

```diff
@@ -46,6 +45,40 @@ def is_library_function(vw, va):
     return vw.funcmeta.get(va, {}).get(_LIBRARY_META_KEY, False)
 
 
+def is_library_called_function(vw, va):
+    """
+    is the given function only called from library functions?
+    if there's no function at the given address, `False` is returned.
+
+    args:
+        vw (vivisect.Workspace):
+        va (int): the virtual address of a function.
+
+    returns:
+        bool: True if the function is only called from library functions, False otherwise.
+    """
```

Comment:
right now, I think this routine is missing the case where a function is deeply nested behind a known library function. For example, A calls B calls C, and A is known to be a library function, therefore we should be able to mark C as well, not just B. So, probably some recursion is necessary.

Reply:
I'd also recommend developing this algorithm alongside some test cases that you've manually verified.

This routine has the potential to dramatically impact capa matching, so we want to be sure we deeply understand how this works and can prove it.

Reply:
> right now, I think this routine is missing the case where a function is deeply nested behind a known library function. For example, A calls B calls C, and A is known to be a library function, therefore we should be able to mark C as well, not just B. So, probably some recursion is necessary.

Great point! I blindly missed this scenario. If I understand correctly, suppose we have: 
```
A (library)
|
v
B
|
v
C
```
Though C isn't directly called by a library function, it's still part of the library's behavior. my initial proposal `is_library_called_function` wouldn't identify C as library-related because it only looked one level deep.

Reply:
@mr-tz @williballenthin I drafted a revision in eba6135ac82ad2052c25db8fe64e39ab4346c49e.

Reply:
is this still untested?

Reply:
@mr-tz,
```
A (library)
|
v
B
|
v
C
```
[`test_function_with_only_top_level_library_functions`](https://github.com/williballenthin/viv-utils/pull/153/files#diff-b6ed29c2c1db14476000efb1b8eba16808854a1e7d913bcf7c97bfab56fc3810R35-R52) should be enough to cover this scenario?

Reply:
it's not easy to see which functions calls which function, can you add comments to elaborate in the test, then it may be good

Reply:
Sure, I will add some code comments to explain the callgraph used in each testcase.